src/ui/FacetValueSuggestions/FacetValueSuggestions.ts

```diff
@@ -0,0 +1,350 @@
+    const preciseResults = rankedResults.splice(0, Math.ceil(this.options.numberOfSuggestions / 2));
+    const broadResults = rankedResults.slice(-1, Math.floor(this.options.numberOfSuggestions / 2));
+
+    return [].concat(preciseResults).concat(broadResults);
+  }
+
+  private getQuerySuggestionsKeywords(omnibox: Omnibox): Promise<string[]> {
+    if (this.options.useQuerySuggestions && omnibox.suggestionAddon) {
+      const suggestions = omnibox.suggestionAddon.getSuggestion();
+      return suggestions ? suggestions.then(suggestions => suggestions.map(s => s.text)) : Promise.resolve();
+    } else {
+      return Promise.resolve();
+    }
+  }
+
+  private computeScoreForSuggestionRow(fieldValue: IIndexFieldValue, reference: IFacetValueReference): IFacetValueSuggestionScore {
+    const totalNumberForFieldValue = reference.fieldsTotal[fieldValue.value] || reference.smallestTotal;
+    const distanceFromTotalForField: number = (totalNumberForFieldValue - fieldValue.numberOfResults) / totalNumberForFieldValue * 100;
+    return {
+      distanceFromTotalForField: distanceFromTotalForField
+    };
+  }
+
+  private listFieldValuesBatch(valuesToSearch: string[]): Promise<IFacetValueSuggestionsResponse> {
+    // The reference request will be used to get the maximum number of values for a given facet value.
+    const referenceValuesRequest = this.buildReferenceFieldValueRequest();
+    const queryParts = this.getQueryToExecuteParts();
+    const suggestionValuesRequests = valuesToSearch.map(value => this.buildListFieldValueRequest(queryParts.concat(value).join(' ')));
+    const requests = [].concat(suggestionValuesRequests).concat(referenceValuesRequest);
+    return this.queryController
+      .getEndpoint()
+      .listFieldValuesBatch({
+        batch: requests
+      })
+      .then(values => {
+        const reference = this.computeReferenceFromBatch(values.pop());
+        const remainingResponses: IFacetValueBatchResponse[] = values.map((value, i) => {
+          return <IFacetValueBatchResponse>{
+            keyword: valuesToSearch[i],
+            values: value
+          };
+        });
+        return <IFacetValueSuggestionsResponse>{
+          responses: remainingResponses,
+          reference: reference
+        };
+      });
+  }
+
+  private computeReferenceFromBatch(batch: IIndexFieldValue[]): IFacetValueReference {
+    const reference: IFacetValueReference = {
+      fieldsTotal: {},
+      smallestTotal: batch[batch.length - 1].numberOfResults
+    };
+    batch.forEach(value => (reference.fieldsTotal[value.value] = value.numberOfResults));
+    return reference;
+  }
+
+  private onRowSelection(row: IFacetValueSuggestionRow, omnibox: Omnibox): void {
+    omnibox.setText(row.keyword);
+    /*args.closeOmnibox();
+    args.clearSuggestions();*/
+    this.queryStateModel.set(this.queryStateFieldValueId, [row.value]);
+    this.usageAnalytics.logSearchEvent<IAnalyticsNoMeta>(analyticsActionCauseList.omniboxField, {});
```

Comment:
We'd probably want to create a new type of analytics event for this component precisely. So this means adding stuff in `AnalyticsActionListMeta.ts`

Also, for ML, there's a lot of stuff that we log related to query suggestions so that ML can improve over time. This include the "partial queries" that we did, the suggestions that we've shown, the ranking of the suggestion that we selected, and so on.

Since this does replace standard query suggestions, the same logic will need to be applied here so that ML continues working properly and improve over time.

This does complicate things a bit, as we'll need to refactor how it works currently, since the code is pretty complex already in the Omnibox component for this.